### DISCUSSION TITLE
Is there a way to parse object-type params?

### DISCUSSION THREAD

**Discussion by geotre** [c-blake/cligen#204] [Q&A] [answered]
I had a look through the documentation and examples but couldn't see anything relevant. In my application I want to have a list of actions to be performed by one of the commands, where each action is represented as something like:

```nim
type Action = object
  name, cmd, args: string
```
And the proc passed to cligen's dispatch is something like:

```nim
proc runActions(a: seq[Action]) = discard

import cligen
dispatch(runActions)
```

In practice these will probably be written out into a config file and merged in using `parsecfg`.

Is this possible? Should I just have a separate config file and handle all this separately to cligen?

**Comment by c-blake** [comment, marked as answer]
I do mention it in the README, but `test/InitOb.nim` may also be helpful to get started.  The essential logic of `initFromCL` is to generate an object initializer from any command-line like data (`seq[string]`, basically).  I'd bet you can adapt that to your purposes.

If you want a *really* worked out example you could look at how [lc](https://github.com/c-blake/lc) uses `cligen`. `lc` actually merges settings from config files, the command line *and* dynamically from per-listed-directory saved dot files called ".lc" (it has to scan a directory anyway - it just loads that if it notices it).  This latter feature allows Unix to be sort of smart like Windows remembering the last "format" for folder contents..It's more manual but also more powerful.  I was even careful to make it workable on file hierarchies where the `lc`-ing user has no write perms through a shadow hierarchy that they *can* write to.  This also allows different users to have different preferences on a shared system.  I'm not quite sure why `lc` is not more popular.  Probably just my not really marketing it as successfully as the RESF markets the not-as-capable IMO `exa`.

**Comment by c-blake** [comment]
Also, if it is ok to be a single command line argument but maybe with commas (,), colons (:), or something, then you can define your own whole new on-the-CLI syntax with an (in-scope of `dispatch`) definition of `proc argParse(dst: var Action, dfl: Action, a: var ArgcvtParams): bool` and `proc argHelp(dfl: Action, a: var ArgcvtParams): seq[string]`.  That is under the Custom Parameter Types part of the README.  You may need to "rotate" your conceptualization slightly, but between these two things, I suspect `cligen` can already do very close to what you want.
